server/app/controllers/admin/AdminImportController.java

```diff
@@ -0,0 +1,145 @@
+   *
+   * <p>This likely could be even smaller.
+   */
+  public static final int MAX_FILE_SIZE_MB = 1;
+
+  private static final Logger logger = LoggerFactory.getLogger(AdminImportController.class);
+
+  private final AdminImportView adminImportView;
+  private final MessagesApi messagesApi;
+
+  private final SettingsManifest settingsManifest;
+
+  @Inject
+  public AdminImportController(
+      AdminImportView adminImportView,
+      MessagesApi messagesApi,
+      ProfileUtils profileUtils,
+      SettingsManifest settingsManifest,
+      VersionRepository versionRepository) {
+    super(profileUtils, versionRepository);
+    this.adminImportView = checkNotNull(adminImportView);
+    this.messagesApi = checkNotNull(messagesApi);
+    this.settingsManifest = checkNotNull(settingsManifest);
+  }
+
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result index(Http.Request request) {
+    if (!settingsManifest.getProgramMigrationEnabled(request)) {
+      return notFound("Program import is not enabled");
+    }
+    return ok(adminImportView.render(request, /* programData= */ Optional.empty()));
+  }
+
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result importProgram(Http.Request request) {
+    if (!settingsManifest.getProgramMigrationEnabled(request)) {
+      return notFound("Program import is not enabled");
+    }
+
+    // See https://www.playframework.com/documentation/2.9.x/JavaFileUpload for how to parse files
+    // in Play Framework.
+    Http.MultipartFormData<Files.TemporaryFile> body = request.body().asMultipartFormData();
+    if (body == null) {
+      return ok(
+          adminImportView.render(request, createErrorResult("Request did not contain a file.")));
+    }
+
+    // The key here must match the `<input type="file">` name set in {@link
+    // views.admin.migration.AdminImportView}.
+    Http.MultipartFormData.FilePart<Files.TemporaryFile> uploadedFile =
+        body.getFile(/* key= */ "file");
+    if (uploadedFile == null) {
+      return ok(
+          adminImportView.render(request, createErrorResult("Request did not contain a file.")));
+    }
+    if (uploadedFile.getFileSize() > MAX_FILE_SIZE_MB * 1024 * 1024) {
+      String error =
+          messagesApi
+              .preferred(request)
+              .at(MessageKey.FILEUPLOAD_VALIDATION_FILE_TOO_LARGE.getKeyName(), MAX_FILE_SIZE_MB);
+      return ok(adminImportView.render(request, createErrorResult(error)));
+    }
+
+    // TODO(#7087): Enable automatic cleaning up of temporary files:
+    // https://www.playframework.com/documentation/2.9.x/JavaFileUpload#Cleaning-up-temporary-files.
+    Files.TemporaryFile uploadedFileRef = uploadedFile.getRef();
+
+    // TODO(#7087): Should this path be defined in application.conf so different deployments can use
+    // different values? Is /tmp a reasonable default?
+    File file = new File("/tmp/imported_program.json");
```

Comment:
The only methods I see on the `TemporaryFile` interface are copying or moving to a file or path: https://www.playframework.com/documentation/2.9.x/api/java/play/libs/Files.TemporaryFile.html, so I don't see any way to directly read the contents of a `TemporaryFile` without creating a `File`. But there might be something I'm missing.

Reply:
Oh actually, uploading it with ajax could solve your form reloading problem.

Reply:
Oh I can give direct file upload a try! Though, that still creates a `File` so I'm not sure it would help with the original question of not creating a `File ` at all. But it would help with the form reloading problem. Let me give it a go.

Reply:
Updated to just use a text box, so resolving.